### PR TITLE
3523 purge bookmarks solr

### LINKED ISSUE #3523: Purge Bookmarks Pointing at Solr Documents that No Longer exist.
### Implementation notes

- Before deleting any bookmarks, ensure that the solr id is not a voyager ID that could be mapped to a valid alma id.
- Be mindful of the number of solr calls you make.

### PR DESCRIPTION
**Pull request by maxkadel**
Closes #3523

### CODE REVIEW

File: spec/factories/bookmarks.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
```
Comment (sandbergja) [inline spec/factories/bookmarks.rb:3]: Nice, thanks for adding this factory!

**Review by sandbergja — CHANGES_REQUESTED** [review]
Thanks, @maxkadel !  It's looking good, but I got an error when I ran `bundle exec rake orangelight:clean:bookmarks:without_solr_documents` on staging:

```
RSolr::Error::Http: RSolr::Error::Http - 414 Request-URI Too Long
Error: <html>
<head><title>414 Request-URI Too Large</title></head>
<body>
<center><h1>414 Request-URI Too Large</h1></center>
<hr><center>nginx/1.23.2</center>
</body>
</html>
```

**Comment by maxkadel** [comment]
> Thanks, @maxkadel ! It's looking good, but I got an error when I ran `bundle exec rake orangelight:clean:bookmarks:without_solr_documents` on staging:
> 
> ```
> RSolr::Error::Http: RSolr::Error::Http - 414 Request-URI Too Long
> Error: <html>
> <head><title>414 Request-URI Too Large</title></head>
> <body>
> <center><h1>414 Request-URI Too Large</h1></center>
> <hr><center>nginx/1.23.2</center>
> </body>
> </html>
> ```

Ok, thanks for testing it (I didn't want to without having had another set of eyes on it). 

It's probably that the default for finding in batches is 1,000, and that's probably too many for building the Solr url. I think we can lower the size of the batches - do you have a guess what size would be reasonable?

File: spec/models/bookmark_spec.rb [hunk 2]
```diff
@@ -7,4 +7,65 @@
...
+    it 'does not change Bookmarks with an alma document_id' do
+      expect do
+        Bookmark.update_to_alma_ids
+      end.not_to change { bookmark_three.reload.document_id }
+    end
+
+    it 'updates bookmarks with a voyager document_id' do
+      expect do
+        Bookmark.update_to_alma_ids
+      end.to change { bookmark_one.reload.document_id }
```
[the remark this answers is not in the document]
Comment (maxkadel) [reply spec/models/bookmark_spec.rb]: Done!

**Comment by maxkadel** [comment]
Hrm, when I ran it on staging, the number of bookmarks went from 2,193 to 110, which makes me a little nervous - that seems like an awfully big reduction. I wonder if it's not waiting long enough for the Solr response / requesting too often?

**Comment by maxkadel** [comment]
> Hrm, when I ran it on staging, the number of bookmarks went from 2,193 to 110, which makes me a little nervous - that seems like an awfully big reduction. I wonder if it's not waiting long enough for the Solr response / requesting too often?

The Solr query was defaulting to only 10 responses, no matter how many ids we requested, so it was calculating way more missing document_ids than were correct. I added a configuration to match the number of bookmarks we look up to the number of rows in the Solr response.

File: spec/models/bookmark_spec.rb [hunk 3]
```diff
@@ -7,4 +7,65 @@
...
+    it 'does not change Bookmarks with an alma document_id' do
+      expect do
+        Bookmark.update_to_alma_ids
+      end.not_to change { bookmark_three.reload.document_id }
+    end
+
+    it 'updates bookmarks with a voyager document_id' do
+      expect do
+        Bookmark.update_to_alma_ids
+      end.to change { bookmark_one.reload.document_id }.from(voyager_id_one).to(converted_voyager_id)
```
Comment (sandbergja) [inline spec/models/bookmark_spec.rb:68]: Cool!  I didn't know about this syntax.

File: app/models/bookmark.rb [hunk 4]
```diff
@@ -12,4 +12,47 @@ class Bookmark < ApplicationRecord
...
+      if doc_ids_without_solr_doc.present?
+        doc_ids_without_solr_doc.each do |doc_id|
+          Bookmark.where(document_id: doc_id)&.destroy_all
+        end
+      end
+    end
+  end
+
+  def self.update_to_alma_ids
+    Bookmark.where("LENGTH(document_id) <= 7").find_in_batches.with_index do |bookmarks, batch|
```
Comment (sandbergja) [inline app/models/bookmark.rb:32]: For peace of mind, could we also confirm that this won't attempt to convert SCSB-*, coin-*, and dsp* ids to alma ids?  I don't think we have one, but theoretically, a document with id:coin-1 would also appear in these results.

**Review by sandbergja — CHANGES_REQUESTED** [review]
Thanks, @maxkadel !  I really like the configurable batch approach.  Just one more thing to check. 🦄

File: app/models/bookmark.rb (hunk 4 above)
Comment (maxkadel) [reply app/models/bookmark.rb:32]: Thanks, good catch, updated.

**Review by kevinreiss — APPROVED** [review]

**Pull request merged**